src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -6,142 +6,174 @@
-            throw new RuntimeException("Illegal length: " + length);
+            throw new IllegalArgumentException("Illegal length: " + length);
         }
-
         try {
-          do {
-              this.keccak.digest(byte_state, 0, BYTE_HASH_LENGTH);
-              //convert to trits
-              tritsFromBigInt(bigIntFromBytes(byte_state, 0, BYTE_HASH_LENGTH), trit_state, 0, Sponge.HASH_LENGTH);
-
-              //copy with offset
-              trit_state[HASH_LENGTH - 1] = 0;
-              System.arraycopy(trit_state, 0, trits, offset, HASH_LENGTH);
-
-              //calculate hash again
-              for (int i = byte_state.length; i-- > 0; ) {
-
-                  byte_state[i] = (byte) (byte_state[i] ^ 0xFF);
-              }
-              keccak.update(byte_state);
-              offset += HASH_LENGTH;
-
-          } while ((length -= HASH_LENGTH) > 0);
+            for (int pos = offset; pos < offset + length; pos += HASH_LENGTH) {
+                this.keccak.digest(byte_state, 0, BYTE_HASH_LENGTH);
+                //convert to trits
+                tritsFromBigInt(bigIntFromBytes(byte_state, 0, BYTE_HASH_LENGTH), trit_state, 0, Sponge.HASH_LENGTH);
+
+                //copy with offset
+                trit_state[HASH_LENGTH - 1] = 0;
+                System.arraycopy(trit_state, 0, trits, pos, HASH_LENGTH);
+
+                //calculate hash again
+                for (int i = byte_state.length; i-- > 0; ) {
+                    byte_state[i] = (byte) (byte_state[i] ^ 0xFF);
+                }
+                keccak.update(byte_state);
+            }
         } catch (DigestException e) {
-            e.printStackTrace();
+            e.printStackTrace(System.err);
             throw new RuntimeException(e);
         }
     }
 
     public static BigInteger bigIntFromTrits(final int[] trits, final int offset, final int size) {
-
+        for (int i = offset; i < offset + size; i++) {
+            if (trits[i] < -1 || trits[i] > 1) {
+                throw new IllegalArgumentException("not a trit: " + trits[i]);
+            }
+        }
         BigInteger value = BigInteger.ZERO;
-
-        for (int i = size; i-- > 0; ) {
-
-            value = value.multiply(BigInteger.valueOf(Converter.RADIX)).add(BigInteger.valueOf(trits[offset + i]));
+        long num;
+        int power = 0;
+        int n = offset + size - 1;
+        for (; n > offset + 1; n--) {
+            switch (trits[n]) {
+                case 0: // a * b
+                    power++;
+                    break;
+                case 1: // a * b + 1
+                    num = 9 + (trits[--n] * 3) + trits[--n];
+                    for (int count = 0; n > offset && count <= 36; count++, power++) {
+                        num = 3 * num + trits[--n];
+                    }
+                    value = value.multiply(RADIX_POWERS[power + 3]).add(BigInteger.valueOf(num));
+                    power = 0;
+                    break;
```

Comment:
Up to 36 trits can be safely processed without using a BigInteger for the math. 
This speeds things up considerably.